File: project/src/services/export.ts
import { PlotStructure, Character, Document, Theme } from '../contexts/AppContext';
import { invoke } from '@tauri-apps/api/core';

export interface ExportOptions {
  format: 'pdf' | 'docx' | 'txt' | 'json' | 'markdown' | 'html' | 'epub';
  includeBeats: boolean;
  includeCharacters: boolean;
  includeThemes: boolean;
  includeConflicts: boolean;
  includeVisualizations: boolean;
  title?: string;
  author?: string;
  description?: string;
}

// Types matching Rust backend
interface RustExportOptions {
  format: string;
  include_beats: boolean;
  include_characters: boolean;
  include_themes: boolean;
  include_conflicts: boolean;
  include_visualizations: boolean;
  title?: string;
  author?: string;
  description?: string;
}

interface RustProjectData {
  title: string;
  author?: string;
  description?: string;
  plot_structure: any;
  characters: any[];
  documents: any[];
  themes: any[];
}

interface ExportResult {
  success: boolean;
  file_path?: string;
  content?: string;
  error?: string;
}

class ExportService {
  async exportPlotStructure(
    plotStructure: PlotStructure,
    characters: Character[],
    documents: Document[],
    themes: Theme[] = [],
    options: ExportOptions
  ): Promise<ExportResult> {
    try {
      // Use Rust backend for all export operations
      const rustOptions: RustExportOptions = {
        format: options.format,
        include_beats: options.includeBeats,
        include_characters: options.includeCharacters,
        include_themes: options.includeThemes,
        include_conflicts: options.includeConflicts,
        include_visualizations: options.includeVisualizations,
        title: options.title,
        author: options.author,
        description: options.description
      };

      const projectData: RustProjectData = {
        title: options.title || 'Untitled Project',
        author: options.author,
        description: options.description,
        plot_structure: plotStructure,
        characters: characters,
        documents: documents,
        themes: themes
      };

      const result: ExportResult = await invoke('export_project', {
        projectData,
        options: rustOptions
      });

      return result;
    } catch (error) {
      console.error('Export failed:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Export failed'
      };
    }
  }

  async exportBeatSheet(
    plotStructure: PlotStructure,
    options: ExportOptions
  ): Promise<ExportResult> {
    try {
      const rustOptions: RustExportOptions = {
        format: options.format,
        include_beats: options.includeBeats,
        include_characters: options.includeCharacters,
        include_themes: options.includeThemes,
        include_conflicts: options.includeConflicts,
        include_visualizations: options.includeVisualizations,
        title: options.title,
        author: options.author,
        description: options.description
      };

      const result: ExportResult = await invoke('export_beat_sheet', {
        plotStructure,
        options: rustOptions
      });

      return result;
    } catch (error) {
      console.error('Beat sheet export failed:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Beat sheet export failed'
      };
    }
  }

  async exportDocument(
    document: Document,
    options: Omit<ExportOptions, 'includeBeats' | 'includeCharacters' | 'includeThemes' | 'includeConflicts' | 'includeVisualizations'>
  ): Promise<ExportResult> {
    try {
      const rustOptions = {
        format: options.format,
        title: options.title || document.title,
        author: options.author,
        description: options.description
      };

      // For simple document export, we can create a minimal project structure
      const projectData = {
        title: document.title,
        author: options.author,
        description: options.description,
        documents: [document],
        plot_structure: null,
        characters: [],
        themes: []
      };

      const result: ExportResult = await invoke('export_project', {
        projectData,
        options: rustOptions
      });

      return result;
    } catch (error) {
      console.error('Document export failed:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Document export failed'
      };
    }
  }

  // Helper method to get supported formats
  getSupportedFormats(): string[] {
    return ['pdf', 'docx', 'txt', 'json', 'markdown', 'html', 'epub'];
  }

  // Helper method to validate export options
  validateExportOptions(options: ExportOptions): { valid: boolean; errors: string[] } {
    const errors: string[] = [];
    
    if (!this.getSupportedFormats().includes(options.format)) {
      errors.push(`Unsupported format: ${options.format}`);
    }

    if (!options.includeBeats && !options.includeCharacters && !options.includeThemes && !options.includeConflicts) {
      errors.push('At least one content type must be included in the export');
    }

    return {
      valid: errors.length === 0,
      errors
    };
  }
}

export const exportService = new ExportService();
export { ExportService };
export type { ExportResult };

File: project/src-tauri/src/commands/export.rs

//! Export-related Tauri commands.

use crate::export::{ExportService, ExportOptions, ProjectData, ExportResult};
use crate::commands::state::AppState;
use tauri::State;

/// Export project data to various formats
#[tauri::command]
pub async fn export_project(
    state: State<'_, AppState>,
    project_data: ProjectData,
    options: ExportOptions,
) -> Result<ExportResult, String> {
    state
        .services()
        .export_service()
        .export_project(&project_data, &options)
        .await
        .map_err(|e| e.to_string())
}

/// Export beat sheet as standalone document
#[tauri::command]
pub async fn export_beat_sheet(
    state: State<'_, AppState>,
    plot_structure: crate::export::PlotStructure,
    options: ExportOptions,
) -> Result<ExportResult, String> {
    state
        .services()
        .export_service()
        .export_beat_sheet(&plot_structure, &options)
        .await
        .map_err(|e| e.to_string())
}

/// Export character profiles
#[tauri::command]
pub async fn export_characters(
    state: State<'_, AppState>,
    characters: Vec<crate::ai::Character>,
    options: ExportOptions,
) -> Result<ExportResult, String> {
    // Create a simplified project data with just characters
    let project_data = ProjectData {
        id: "characters-export".to_string(),
        name: "Character Profiles".to_string(),
        description: "Exported character profiles".to_string(),
        documents: vec![],
        characters,
        plot_structure: None,
        notes: vec![],
        created_at: chrono::Utc::now(),
        updated_at: chrono::Utc::now(),
    };

    let export_options = ExportOptions {
        include_metadata: false,
        include_plot_structure: false,
        include_characters: true,
        include_notes: false,
        ..options
    };

    state
        .services()
        .export_service()
        .export_project(&project_data, &export_options)
        .await
        .map_err(|e| e.to_string())
}

/// Export research notes
#[tauri::command]
pub async fn export_research_notes(
    state: State<'_, AppState>,
    notes: Vec<crate::export::Note>,
    options: ExportOptions,
) -> Result<ExportResult, String> {
    // Create a simplified project data with just notes
    let project_data = ProjectData {
        id: "research-export".to_string(),
        name: "Research Notes".to_string(),
        description: "Exported research notes".to_string(),
        documents: vec![],
        characters: vec![],
        plot_structure: None,
        notes,
        created_at: chrono::Utc::now(),
        updated_at: chrono::Utc::now(),
    };

    let export_options = ExportOptions {
        include_metadata: false,
        include_plot_structure: false,
        include_characters: false,
        include_notes: true,
        ..options
    };

    state
        .services()
        .export_service()
        .export_project(&project_data, &export_options)
        .await
        .map_err(|e| e.to_string())
}
